src/corppa/ocr/gvision_ocr.py

```diff
@@ -7,60 +7,168 @@
+    )
+    sys.exit(1)
+
+
+def image_relpath_generator(image_dir, ext_set, follow_symlinks=True):
+    for dirpath, dirs, files in image_dir.walk(follow_symlinks=follow_symlinks):
+        # Check the files in walked directory
+        for file in files:
+            ext = os.path.splitext(file)[1]
+            if ext:
+                filepath = dirpath.joinpath(file)
+                yield filepath.relative_to(image_dir)
+        # For future walking, remove hidden directories
+        dirs[:] = [d for d in dirs if d[0] != "."]
+
+
+def get_image_ocr(in_dir, out_dir, ext_list, ocr_limit=0, show_progress=True):
     # Instantiate google vision client
     client = vision.ImageAnnotatorClient()
 
-    for imagefile in glob.iglob("%s/*/*.jpg" % IMAGE_DIR):
-        basepath = os.path.splitext(imagefile)[0]
-        textfile = "%s.txt" % basepath
-        jsonfile = "%s.json" % basepath
-
-        # if text file already exists, skip
-        if os.path.isfile(textfile):
-            continue
-
-        # if text file does not exist, request ocr
-
-        # Load the image into memory
-        with io.open(imagefile, "rb") as image_file:
-            content = image_file.read()
-        image = vision.Image(content=content)
-
-        # Performs ocr and handwriting detection on the image file
-        response = client.document_text_detection(image=image)
-
-        # save plain text output to local file;
-        # even if text is empty, create text file so we don't request again
-        with open(textfile, "w") as textfilehandle:
-            textfilehandle.write(response.full_text_annotation.text)
-
-        # this generates a json string
-        json_response = vision.AnnotateImageResponse.to_json(response)
-        with open(jsonfile, "w") as jsonfilehandle:
-            jsonfilehandle.write(json_response)
+    # Setup up progress bar if progress will be shown
+    if show_progress:
+        desc = "OCRing images"
+        maxinterval = 1
+        if ocr_limit:
+            progress_bar = tqdm(desc=desc, total=ocr_limit, maxinterval=maxinterval)
+        else:
+            bar_format = "{desc}: {n:,} images OCR'd | elapsed: {elapsed}, {rate_fmt}"
+            progress_bar = tqdm(
+                desc=desc, bar_format=bar_format, maxinterval=maxinterval
+            )
 
-        if response.error.message:
-            raise Exception(
-                "{}\nFor more info on error messages, check: "
-                "https://cloud.google.com/apis/design/errors".format(
-                    response.error.message
+    ocr_count = 0
+    for image_relpath in image_relpath_generator(in_dir, set(ext_list)):
+        # Refresh progress bar
+        if show_progress:
+            progress_bar.refresh()
+        imagefile = in_dir.joinpath(image_relpath)
+        textfile = out_dir.joinpath(image_relpath).with_suffix(".txt")
+        jsonfile = out_dir.joinpath(image_relpath).with_suffix(".json")
+
+        # Request OCR if file does not exist
```

Comment:
starting here looks like roughly the block of logic to pull out into a separate function that you can call from either mode of running this script (specific volumes or not)

Reply:
Oh, oops. I think we were thinking about how we would tackle this. Hopefully the latest revisions will address this sufficiently.